flecsi/data/hpx/policy.hh

```diff
@@ -0,0 +1,418 @@
+    // ranks and indices of the destination of copy i.e. (local source
+    // index, {(remote dest rank, remote dest index)}). We need to do a shuffle
+    // operation to reconstruct this info from {(local ghost index, remote
+    // source rank, remote source index)}.
+    auto remote_sources =
+      destination.get_storage<const points::Value>(meta_fid);
+    // Essentially a GroupByKey of remote_sources, keys are the remote source
+    // ranks and values are vectors of remote source indices.
+    SendPoints remote_shared_entities;
+    for(const auto & [begin, end] : destination.ghost_ranges) {
+      for(auto ghost_idx = begin; ghost_idx < end; ++ghost_idx) {
+        const auto & shared = remote_sources[ghost_idx];
+        remote_shared_entities[shared.first].emplace_back(shared.second);
+        // We also group local ghost entities into
+        // (src rank, { local ghost ids})
+        ghost_entities[shared.first].emplace_back(ghost_idx);
+      }
+    }
+
+    // Create the inverse mapping of
+    // group_shared_entities. This creates a map from remote destination rank to
+    // a vector of *local* source indices. This information is later used by
+    // hpx::collectives::set().
+    {
+      auto comm = flecsi::run::context::instance().world_comm(
+        "copy_engine" + std::to_string(meta_fid));
+      auto f = util::hpx::all_to_allv(
+        [&](int r, int) -> auto & {
+          static const std::vector<std::size_t> empty;
+          const auto i = remote_shared_entities.find(r);
+          return i == remote_shared_entities.end() ? empty : i->second;
+        },
+        comm);
+
+      std::size_t r = 0;
+      for(auto & v : f.get()) {
+        if(!v.empty())
+          shared_entities.try_emplace(r, std::move(v));
+        ++r;
+      }
+    }
+
+    // We need to figure out the max local source index in order to give correct
+    // nelems when calling region::get_storage().
+    for(const auto & [rank, indices] : shared_entities) {
+      max_local_source_idx = (std::max)(max_local_source_idx,
+        *std::max_element(indices.begin(), indices.end()));
+    }
+    max_local_source_idx += 1;
+  }
+
+  // called with each field (and field_id_t) on the entity, for example, one
+  // for pressure, temperature, density etc.
+  void operator()(field_id_t data_fid) const {
+
+    auto source_storage =
+      source.r->get_storage<std::byte>(data_fid, max_local_source_idx);
+    auto destination_storage = destination.get_storage<std::byte>(data_fid);
+    auto type_size = source.r->get_field_info(data_fid)->type_size;
+
+    auto comm = flecsi::run::context::instance().world_channel_comm();
+
+    using buffer_type = ::hpx::serialization::serialize_buffer<std::byte>;
+    using data_type = std::vector<buffer_type>;
+
+    std::vector<::hpx::future<void>> requests;
+    requests.reserve(ghost_entities.size() + shared_entities.size());
+
+    for(const auto & [src_rank, ghost_indices] : ghost_entities) {
```

Comment:
Does this block of the code implement  ghost-copy? I am just trying to understand the code

Reply:
@ipdemes hmmm, I'm not sure what code you are referring to.  All send/receive operations are asynchronous and overlapping. The code synchronizes by waiting for the futures further down (here: https://github.com/flecsi/flecsi/pull/1/files/7c97f33247a4fe9eacebc41b831194e4613564d2#diff-7b8feaa88283f1c8911d0447df01a47f7222d08b55679c0b210fcdb7da71b070R394).

Reply:
@ipdemes 

Most of this file came from the MPI backend with MPI_Send/Recv replaced with hpx::collective::get/set and "asynchronized".

Reply:
@ipdemes 

To answer the question if this block of code does ghost copy, it does. It is equivalent to issuing MPI_IRecv on the destination side of the ghost copy for the "Aggregated" message. The continuation in the .then() does the split/scattering of the aggregated message into their final destination in the storage.